### PR TITLE
Improve render minicart content by native binding knockoutJs

### LINKED ISSUE #30564: [Issue] Improve render minicart content by native binding knockoutJs
This issue is automatically created based on existing pull request: magento/magento2#30148: Improve render minicart content by native binding knockoutJs

---------

### Preconditions: (*)

Magento 2.4-develop

This PR aim to bring improvement for render minicart content by default knockout binding instead use magento binding. Default magento provide it own custom binding and used widely across core base
More details here: https://devdocs.magento.com/guides/v2.4/ui_comp_guide/concepts/magento-bindings.html
With new patch minicart content will render faster
We should delivery content as fast as possible for user at first load. Improve user experiences

### Steps to reproduce: (*)

1. Test in some main page such as : Home , Category, Details Page
2. Add product to cart
3. Verify minicart content update and render without break
For profiler deep check we need open "Performance" tab in google chrome dev tools
Reload from start record for profiling
Zoom in graph profile > See Network section and find and indicate
Magento/Checkout/view/frontend/web/template/minicart/content.html rendered
See Main section to view how much time  parseTemplate called

### Questions or comments

Why choose minicart? Because it's show in almost page and render while user interactive to ui

**Actual Result:** :heavy_multiplication_x: 
before_update
Time parse: ~7ms

**Expected Result:** :heavy_check_mark: 
after_update
Time parse: ~2ms

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by mrtuvn**
### Description (*)

This PR aim to bring improvement for render minicart content by default knockout binding instead use magento binding. Default magento provide it own custom binding and used widely across core base
More details here: https://devdocs.magento.com/guides/v2.4/ui_comp_guide/concepts/magento-bindings.html
With new patch minicart content will render faster
We should delivery content as fast as possible for user at first load. Improve user experiences

### Fixed Issues (if relevant)

No issue available related 

### Manual testing scenarios (*)

1. Test in some main page such as : Home , Category, Details Page
2. Add product to cart
3. Verify minicart content update and render without break
For profiler deep check we need open "Performance" tab in google chrome dev tools
Reload from start record for profiling
Zoom in graph profile > See Network section and find and indicate
Magento/Checkout/view/frontend/web/template/minicart/content.html rendered
See Main section to view how much time  parseTemplate called

Test check qty visual in minicart
when use config Display Cart Summary: Display item quantities <=Default option
Admin > Configuration > Sales > Checkout > My Cart Link

Test case 1: Add only one product and update qty to above a least 2
Expected result: show correct item qty 
Test case 2: Add 2 products and update qty
Expected result show correct item qty 

### Questions or comments

Why choose minicart? Because it's show in almost page and render while user interactive to ui

Before change results
before_update
Time parse: ~7ms

After change results
after_update
Time parse: ~2ms

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#30564: Improve render minicart content by native binding knockoutJs

### CODE REVIEW

**Comment by kandy** [comment]
@mrtuvn Can you describe problem that you want to resolve by this PR?

**Comment by mrtuvn** [comment]
@kandy Just rewrite minicart template use native ko binding for imrpove better performance render html
Can we use any tools for measure this changes

**Comment by ihor-sviziev** [comment]
@mrtuvn i thought it should have approximately equal performance and shouldn’t be really noticeable. It’s very strange...

**Comment by omiroshnichenko** [comment]
Hi @mrtuvn, I completely agree with @ihor-sviziev. Could you please make local performance measurement using Chrome develop tools and share results on how it improves performance?

**Review by sidolov — CHANGES_REQUESTED** [review]
Waiting for performance measurements

**Comment by mrtuvn** [comment]
Tested with product "Chaz Kangeroo Hoodie" SKU MH01
all cache enabled varnish 6 + redis 5 + Use chrome 85.0.4183.102  + Ubuntu 20.04.1 arm64
Render before change
before
Render after change
after

**Comment by omiroshnichenko** [comment]
@mrtuvn Thanks for the report. But, the metric that you selected it is not on what we need to take a look at. On screenshots, you are showing the request time, its a time to fetch the file from the browser cache or server. Since the processing of custom tags/nodes is happening on the frontend part(in browser) we need to take a look at how much it took to convert custom nodes to regular knockout syntax.

**Comment by ihor-sviziev** [comment]
@mrtuvn but seems like you highlighted the network request that will not depends on your fix for sure, or I just looking in wrong place?

**Comment by mrtuvn** [comment]
@omiroshnichenko Thanks for your the reply. Previous i just used chrome performance tab for record. But it can affect by some factors and lead to inaccurate results
 Is there any other way to measure it correctly ?

**Comment by mrtuvn** [comment]
@Karlasa @krzksz @jonathanKingston @DrewML can you guys leave some comments from frontend view

**Comment by omiroshnichenko** [comment]
@mrtuvn Continue using the Chrome performance tab. You need to find a section with the conversion of the template and take a look at the execution time.

**Comment by krzksz** [comment]
I have numbers of this, gonna post them ASAP.

**Comment by ihor-sviziev** [comment]
@krzksz any updates ?

**Comment by krzksz** [comment]
Alright sorry for the delay but here we go. So the function responsible for all of the transformations from Magento custom logic to KnockoutJS bindings is `normalize` called by `processTemplate` in `Magento_Ui/js/lib/knockout/template/renderer.js`. Every time HTML template is loaded for a component, it gets parsed and adjusted accordingly so that KnockoutJS actually understands which bindings it should apply.

Here's a before:
obraz
And after applying the changes:
obraz

All in all, the difference is going from almost 50% of component initialization down to ~20-25% with still a room from improvement.

I think I remember there were some guidelines that developers should use Magento's custom bindings instead of `data-bind`, but in this case we seem to be sacrificing a portion of user experience for developer experience and this is not a way to go in my opinion.

File: app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html [hunk 1]
```diff
@@ -4,6 +4,7 @@
  * See COPYING.txt for license details.
  */
 -->
+<!-- Improvement render html minicart content by native binding knockoutJs -->
```
Comment (krzksz) [inline app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html]: This comment is unnecessary and should be removed as the purpose of your changes is already included in the commit message.

File: app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html [hunk 2]
```diff
@@ -30,23 +31,26 @@
...
+            <!-- ko if: maxItemsToDisplay < getCartLineItemsCount() --><!-- ko i18n: 'of' --><!-- /ko --><!-- /ko -->
             <span class="count" text="getCartParam('summary_count')"/>
-                <!-- ko if: (getCartLineItemsCount() === 1) -->
-                    <span translate="'Item in Cart'"/>
-                <!--/ko-->
-                <!-- ko if: (getCartLineItemsCount() > 1) -->
-                    <span translate="'Items in Cart'"/>
-                <!--/ko-->
+            <!-- ko if: (getCartLineItemsCount() === 1) -->
+                <span translate="'Item in Cart'"/>
```
Comment (krzksz) [inline app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html]: I think we should also replace `translate` attribute with `data-bind="i18n: ''Item in Cart''"` and everywhere else if we want to go full Knockout native. Please go through the whole file.

File: app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html [hunk 3]
```diff
@@ -30,23 +31,26 @@
...
+            <!--/ko-->
+            <!-- ko if: (getCartLineItemsCount() > 1) -->
+                <span translate="'Items in Cart'"/>
```
Comment (krzksz) [inline app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html]: Same here.

File: app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html [hunk 4]
```diff
@@ -64,24 +68,25 @@
                 <div data-bind="html: getCartParamUnsanitizedHtml('extra_actions')"></div>
             </div>
         </div>
-    </if>
+        <!-- /ko -->
+    <!-- /ko -->
 
-    <if args="getCartParam('summary_count')">
+    <!-- ko if: getCartParam('summary_count') -->
         <strong class="subtitle" translate="'Recently added item(s)'"/>
```
Comment (krzksz) [inline app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html]: Same here with change to `data-bind="i18n: ''..."`.

File: app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html [hunk 5]
```diff
@@ -64,24 +68,25 @@
...
+                <!-- /ko -->
             </ol>
         </div>
-    </if>
+    <!-- /ko -->
 
-    <ifnot args="getCartParam('summary_count')">
+    <!-- ko ifnot: getCartParam('summary_count') -->
         <strong class="subtitle empty"
                 translate="'You have no items in your shopping cart.'"
```
Comment (krzksz) [inline app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html]: Same here with change to `data-bind="i18n: ''..."`.

File: app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html [hunk 6]
```diff
@@ -64,24 +68,25 @@
...
         />
-        <if args="getCartParam('cart_empty_message')">
+        <!-- ko if: getCartParam('cart_empty_message') -->
             <p class="minicart empty text" text="getCartParam('cart_empty_message')"/>
```
Comment (krzksz) [inline app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html]: This should also be changed to Knockout's binding using `data-bind="text: '...'"`

File: app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html [hunk 7]
```diff
@@ -90,19 +95,23 @@
...
-    </ifnot>
+        <!-- /ko -->
+    <!-- /ko -->
 
-    <div class="actions" if="getCartParam('summary_count')">
+    <!-- ko if: getCartParam('summary_count') -->
+    <div class="actions">
         <div class="secondary">
             <a class="action viewcart" data-bind="attr: {href: shoppingCartUrl}">
                 <span translate="'View and Edit Cart'"/>
```
Comment (krzksz) [inline app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html]: Same here with `translate` attribute.

File: app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html [hunk 8]
```diff
@@ -30,23 +31,26 @@
...
-    <if args="getCartParam('summary_count')">
+    <!-- ko if: getCartParam('summary_count') -->
         <div class="items-total">
-            <span class="count" if="maxItemsToDisplay < getCartLineItemsCount()" text="maxItemsToDisplay"/>
-            <translate args="'of'" if="maxItemsToDisplay < getCartLineItemsCount()"/>
+            <!-- ko if: maxItemsToDisplay < getCartLineItemsCount() -->
+            <span class="count" text="maxItemsToDisplay"/>
+            <!-- /ko -->
+            <!-- ko if: maxItemsToDisplay < getCartLineItemsCount() --><!-- ko i18n: 'of' --><!-- /ko --><!-- /ko -->
             <span class="count" text="getCartParam('summary_count')"/>
```
Comment (krzksz) [inline app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html]: This can be changed to `data-bind="text: '...'"`.

**Review by krzksz — CHANGES_REQUESTED** [review]

**Comment by krzksz** [comment]
Ok, I added my comments what needs to happen if we want to go full throttle with this approach. The question still stands if we really want to...

**Comment by mrtuvn** [comment]
i have updated branch with mainline code base and also update PR as @krzksz suggestion

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
Here is the docs related with @krzksz perference
https://devdocs.magento.com/guides/v2.4/ui_comp_guide/concepts/magento-bindings.html

> Within HTML templates, Magento gives you the option of using a binding syntax that is simpler and easier to read and write than the standard Knockout binding syntax

**Comment by krzksz** [comment]
Thanks for the changes, the numbers for all of the changes are:

**Before:**
~20ms of ~50ms total
**After:**
~5.5ms of ~35.5ms total

You may think this is not much of a difference, but it is enough to convince Chrome to stop flagging minicart initialization as long running task.

**Comment by krzksz** [comment]
@magento run Functional Tests B2B,Integration Tests

**Review by krzksz — APPROVED** [review]

**Comment by krzksz** [comment]
No functional changes, E2E tests pass so no additional tests are required IMHO.

**Comment by magento-engcom-team** [comment]
Hi @krzksz, thank you for the review.
[ENGCOM-8306](https://jira.corp.magento.com/browse/ENGCOM-8306) has been created to process this Pull Request

**Comment by ihor-sviziev** [comment]
@mrtuvn @krzksz 
Just an idea - maybe let's add some notice to the https://devdocs.magento.com/guides/v2.4/ui_comp_guide/concepts/magento-bindings.html page that it works a big slower than knockout bindings and it's recommended to use knockout bindings for the critical parts, like minicart?

**Comment by mrtuvn** [comment]
> @mrtuvn @krzksz 
> Just an idea - maybe let's add some notice to the https://devdocs.magento.com/guides/v2.4/ui_comp_guide/concepts/magento-bindings.html page that it works a bit slower than knockout bindings and it's recommended to use knockout bindings for the critical parts, like minicart?

I have create or for that

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Test in some main page such as : Home, Category, Details Page;
2. Add product to cart;
3. Verify mini cart content update and render without break;

**Actual Result:** :heavy_check_mark: mini cart updated and rendered without breaks

For profiler deep check we need to open "Performance" tab in google chrome dev tools
4. Reload from start record for profiling
5. Zoom in graph profile > See Network section and find and indicate
Magento/Checkout/view/frontend/web/template/minicart/content.html rendered
6. See Main section to view how much time parseTemplate called

**Before:** :heavy_multiplication_x: Time parse: ~7ms



**After:** :heavy_check_mark:  Time parse: ~2ms

**Comment by engcom-Alfa** [comment]
@magento create issue

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
rebased with mainline code

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by hostep** [comment]
@mrtuvn : was there any specific reason for rebasing your changes? The PR was already approved and tested, so now QA has to start over again I think since it's hard to tell if anything changed in your force pushes 😛  (I might be wrong, but it feels unnecessary to rebase if it's not needed)

File: app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html [hunk 9]
```diff
@@ -27,26 +33,29 @@
...
+            <!-- /ko -->
+            <!-- ko if: maxItemsToDisplay < getCartLineItemsCount() --><!-- ko i18n: 'of' --><!-- /ko --><!-- /ko -->
+            <span class="count" data-bind="text: getCartParam('summary_count')"/>
+            <!-- ko if: (getCartParam('summary_count') === 1) -->
+                <span data-bind="i18n: 'Item in Cart'"/>
+            <!-- /ko -->
+            <!-- ko if: (getCartParam('summary_count') > 1) -->
+                <span data-bind="i18n: 'Items in Cart'"/>
+            <!-- /ko -->
         </div>
```
Comment (mrtuvn) [inline app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html:52]: @hostep i have updated here for correct check items count qty

Comment (mrtuvn) [reply app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html:52]: replace getCartLineItemsCount() by getCartParam('summary_count')

Comment (hostep) [reply app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html:52]: Ah ok, that wasn't really clear to me, you can always leave a comment saying what you've changed. Sorry about the noise! 🙂

**Comment by mrtuvn** [comment]
Update mftf test cover for case MC-14726: Create a simple products with all types of oprtions and add it to the cart

**Comment by mrtuvn** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B, Integration Tests

**Comment by mrtuvn** [comment]
@magento run test Database Compare, Magento Health Index, Sample Data Tests CE, Sample Data Tests EE, Sample Data Tests B2B, Static Tests, Unit Tests, WebAPI Tests, Semantic Version Checker

**Comment by ihor-sviziev** [comment]
@magento run Database Compare

File: app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html (hunk 9 above)
Comment (ihor-sviziev) [reply app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html:52]: > replace getCartLineItemsCount() by getCartParam('summary_count')

Seems like this is a bug fix that should be double checked.

Test failures looks not related to changes from this PR.

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8306](https://jira.corp.magento.com/browse/ENGCOM-8306) has been created to process this Pull Request

File: app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html (hunk 9 above)
Comment (mrtuvn) [reply app/code/Magento/Checkout/view/frontend/web/template/minicart/content.html:52]: @ihor-sviziev i have updated test case for QA

File: app/code/Magento/Checkout/Test/Mftf/Test/StorefrontAddProductWithAllTypesOfCustomOptionToTheShoppingCartTest.xml [hunk 10]
```diff
@@ -105,7 +105,7 @@
         <!-- Assert Product Count in Mini Cart -->
         <actionGroup ref="StorefrontAssertMiniCartItemCountActionGroup" stepKey="assertProductCountAndTextInMiniCart">
             <argument name="productCount" value="2"/>
-            <argument name="productCountText" value="2 Item in Cart"/>
+            <argument name="productCountText" value="2 Items in Cart"/>
         </actionGroup>
```
Comment (mrtuvn) [inline app/code/Magento/Checkout/Test/Mftf/Test/StorefrontAddProductWithAllTypesOfCustomOptionToTheShoppingCartTest.xml:110]: @ihor-sviziev during test i just find out this issue. We can cover and test this case too

Comment (mrtuvn) [reply app/code/Magento/Checkout/Test/Mftf/Test/StorefrontAddProductWithAllTypesOfCustomOptionToTheShoppingCartTest.xml:110]: Only visual issue not actual big problem

Comment (ihor-sviziev) [reply app/code/Magento/Checkout/Test/Mftf/Test/StorefrontAddProductWithAllTypesOfCustomOptionToTheShoppingCartTest.xml:110]: I believe adjusting this MFTF test will be enough, but just wanted to make sure QA team will double check it manually in order to make sure there no regression issues.

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

### Case 1

**Manual testing scenario:**

1. Test in some main page such as : Home, Category, Details Page;
2. Add product to cart;
3. Verify mini cart content update and render without break;

**Actual Result:** :heavy_check_mark: mini cart updated and rendered without breaks

For profiler deep check we need to open "Performance" tab in google chrome dev tools
4. Reload from start record for profiling
5. Zoom in graph profile > See Network section and find and indicate
Magento/Checkout/view/frontend/web/template/minicart/content.html rendered
6. See Main section to view how much time parseTemplate called

**Before:** :heavy_multiplication_x: Time parse: ~7ms



**After:** :heavy_check_mark:  Time parse: ~2ms



 ### Case 2

1. Go to Admin -> Stores -> Configuration -> Sales -> Checkout -> My Cart Link and set:
- **Display Cart Summary** to **Display item quantities**
2. Add only one product and update qty to above a least 2

**Actual Reult:** :heavy_check_mark: show correct item qty

Peek 2020-10-23 11-55

3. Add 2 products and update qty to 10

**Actual Reult:** :heavy_check_mark: show correct item qty

Peek 2020-10-23 11-57

 ### Case 3

1. Go to Admin -> Stores -> Configuration -> Sales -> Checkout -> My Cart Link and set:
- **Display Cart Summary** to **Display number of items in cart**
2. Add only one product and update qty to above a least 2

**Actual Reult:** :heavy_check_mark: show correct item qty (1 Item in Cart)

Peek 2020-10-23 12-01


3. Add another product to the cart

**Actual Reult:** :heavy_check_mark: show correct item qty (2 Items in Cart)

Peek 2020-10-23 12-03

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests EE

**Comment by mrtuvn** [comment]
Can we have new proceed on this PR

**Comment by gabrieldagama** [comment]
Hi @mrtuvn, thanks for contributing!

We've discussed this PR internally, even though it does bring a performance improvement as shown by @krzksz, the Magento binding is an important layer for the platform for several reasons, as for example, being the abstraction layer for the KnockoutJS, making it easier for us on future to make internal changes.

We do acknowledge that there is room for improvement, but we request that the performance improvement be made on the Magento bindings itself instead of not using it.

I will close this PR since we cannot accept such changes.

I hope that makes sense. Thanks.

**Pull request closed without merging**